### PR TITLE
fix: Fix UBSan issue

### PR DESCRIPTION
**Pull request by mvadari**
## High Level Overview of Change

This PR is a quick fix that fixes the UBSan crashes in tests.

### Context of Change

```
/__w/rippled/rippled/src/libxrpl/tx/transactors/Sponsor/SponsorshipTransfer.cpp:485:71: runtime error: negation of 1 cannot be represented in type 'std::uint32_t' (aka 'unsigned int')
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /__w/rippled/rippled/src/libxrpl/tx/transactors/Sponsor/SponsorshipTransfer.cpp:485:71
```

### API Impact

N/A

### CODE REVIEW

File: src/libxrpl/tx/transactors/Sponsor/SponsorshipTransfer.cpp [hunk 1]
```diff
@@ -407,9 +408,9 @@ SponsorshipTransfer::doApply()
         return tefINTERNAL;  // LCOV_EXCL_LINE
 
     auto const setSponsorFieldU32 = [](auto const& sle, auto const& field, auto const& delta) {
-        int32_t const newValue = static_cast<int32_t>(sle->getFieldU32(field)) + delta;
+        int64_t const newValue = sle->getFieldU32(field) + delta;
 
-        if (newValue < 0)
+        if (newValue < 0 || newValue > std::numeric_limits<std::uint32_t>::max())
         {
```
[the remark this answers is not in the document]
Comment (mvadari) [reply src/libxrpl/tx/transactors/Sponsor/SponsorshipTransfer.cpp]: 90f9661

File: src/libxrpl/tx/transactors/Sponsor/SponsorshipTransfer.cpp [hunk 2]
```diff
@@ -407,9 +408,9 @@ SponsorshipTransfer::doApply()
         return tefINTERNAL;  // LCOV_EXCL_LINE
 
     auto const setSponsorFieldU32 = [](auto const& sle, auto const& field, auto const& delta) {
-        int32_t const newValue = static_cast<int32_t>(sle->getFieldU32(field)) + delta;
+        int64_t const newValue = static_cast<int64_t>(sle->getFieldU32(field)) + delta;
 
-        if (newValue < 0)
+        if (newValue < 0 || newValue > std::numeric_limits<std::uint32_t>::max())
```
Comment (oleks-rip) [inline src/libxrpl/tx/transactors/Sponsor/SponsorshipTransfer.cpp]: you must throw, not just return

Comment (oleks-rip) [reply src/libxrpl/tx/transactors/Sponsor/SponsorshipTransfer.cpp]: add more () around every for clang-tidy

Comment (oleks-rip) [reply src/libxrpl/tx/transactors/Sponsor/SponsorshipTransfer.cpp]: Or just copy/paste it from 7542

Comment (mvadari) [reply src/libxrpl/tx/transactors/Sponsor/SponsorshipTransfer.cpp]: Why?

File: src/libxrpl/tx/transactors/Sponsor/SponsorshipTransfer.cpp [hunk 3]
```diff
@@ -379,8 +380,8 @@ reduceReserveCount(
     if (!sponsorSle)
         return tefINTERNAL;  // LCOV_EXCL_LINE
 
-    auto const reserveCount = sponsorSle->getFieldU32(sfReserveCount);
-    int32_t const afterReserveCount = reserveCount + delta;
+    int64_t const reserveCount = sponsorSle->getFieldU32(sfReserveCount);
```
Comment (yinyiqian1) [inline src/libxrpl/tx/transactors/Sponsor/SponsorshipTransfer.cpp]: To prevent future regressions where this gets accidentally changed to `auto` (sometimes clang-tidy may suggest), it's worth encapsulating this calculation in a helper function and calling this helper everywhere if needed. The helper keeps the type safety contained in one place.

```
std::optional<std::uint32_t>
applyCountDelta(std::uint32_t current, std::int64_t delta)
{
    std::int64_t const next = static_cast<std::int64_t>(current) + delta;
    if (next < 0 || next > std::numeric_limits<std::uint32_t>::max())
        return std::nullopt;
    return static_cast<std::uint32_t>(next);
}
```
And let's also add tests for this helper, to test the signed boundary behaviors.
In caller places, we can check if the return value is std::nullopt, and add a UNREACHABLE assertion there and return early.

Comment (mvadari) [reply src/libxrpl/tx/transactors/Sponsor/SponsorshipTransfer.cpp]: UBSan will crash anyways if this is changed

Comment (yinyiqian1) [reply src/libxrpl/tx/transactors/Sponsor/SponsorshipTransfer.cpp]: I still highly recommend encapsulating the type safety logic within a single helper function. Handling this at the code-logic level gives us true safety, which is much more robust than relying on a test crash downstream. And it is used not just once. Any thoughts? @oleks-rip

File: src/libxrpl/tx/transactors/Sponsor/SponsorshipTransfer.cpp (hunk 2 above)
Comment (oleks-rip) [reply src/libxrpl/tx/transactors/Sponsor/SponsorshipTransfer.cpp]: Because it is failed, and Unreachable  just silently ignore it in production.

File: src/libxrpl/tx/transactors/Sponsor/SponsorshipTransfer.cpp (hunk 3 above)
Comment (oleks-rip) [reply src/libxrpl/tx/transactors/Sponsor/SponsorshipTransfer.cpp]: I'm ok with one function. Not agree with unreachable, as it is assert it will not work in production, it should throw/ return tecInternal

File: src/libxrpl/tx/transactors/Sponsor/SponsorshipTransfer.cpp (hunk 2 above)
Comment (mvadari) [reply src/libxrpl/tx/transactors/Sponsor/SponsorshipTransfer.cpp]: ff161f1

File: src/libxrpl/tx/transactors/Sponsor/SponsorshipTransfer.cpp (hunk 3 above)
Comment (mvadari) [reply src/libxrpl/tx/transactors/Sponsor/SponsorshipTransfer.cpp]: d027421

**Review by yinyiqian1 — APPROVED** [review]

**Review by oleks-rip — APPROVED** [review]
